op-node/rollup/derive/frame_queue.go

```diff
@@ -56,6 +61,68 @@ func (fq *FrameQueue) NextFrame(ctx context.Context) (Frame, error) {
 	return ret, nil
 }
 
+func (fq *FrameQueue) loadNextFrames(ctx context.Context) error {
+	// Find more frames if we need to
+	if len(fq.frames) == 0 {
+		data, err := fq.prev.NextData(ctx)
+		if err != nil {
+			return err
+		}
+
+		if frame, err := ParseFrames(data); err == nil {
+			fq.frames = append(fq.frames, frame...)
+		} else {
+			fq.log.Warn("Failed to parse frames", "origin", fq.prev.Origin(), "err", err)
+		}
+	}
+	return nil
+}
+
+func (fq *FrameQueue) prune() {
+	fq.frames = pruneFrameQueue(fq.frames)
+}
+
+// pruneFrameQueue prunes the frame queue to only hold contiguous and ordered
+// frames, conforming to Holocene frame queue rules.
+func pruneFrameQueue(frames []Frame) []Frame {
+	for i := 0; i < len(frames)-1; {
+		current, next := frames[i], frames[i+1]
+		discard := func(shift int) {
```

Comment:
Would `num` be a better name than `shift`? We're deleting that many frames, which does shift anything after that left by that amount but that's less direct to my mind (could just be me...).

Reply:
I think `discard` actually only ever deletes a single frame? It's the frame at index `i+shift`. 

I wonder if a better API might be 

```go 
    discard := func(idx int) {
      frames = append(frames[0:idx], frames[idx+1:]...)
     }
```

and then at the callsite we have e.g. 
`discard(i+1)` // discard next
`discard(i)` // discard current

And the added bonus that we don't have to declare a new function on every loop iteration.

Reply:
I chose `shift` because it's a shift of the current cursor `i`. This doesn't discard `shift` many frames, but the frame at position `i + shift`. Renamed to `d` to avoid wrong connotations.